Support mixed case service ID

CindyZX opened the pull request:
This PR is to address issue https://github.com/zowe/api-layer/issues/215. Prior to this PR, gateway url with mixed case service ID can not be routed. The reason is that, service ID is converted to lower cases in Discovery Service

I modified MfaasRouteLocator.java, recover serviceId by setting it to be VIPAdress. Because when service instance is created, VIPAddress is set to serviceId, and it is not converted to lower case by Discovery Service.
```
     builder.setAppName(serviceId).setInstanceId(instanceId).setHostName(url.getHost()).setIPAddr(ipAddress)
            .setDataCenterInfo(DEFAULT_INFO).setVIPAddress(serviceId).setSecureVIPAddress(serviceId)
            .setLeaseInfo(LeaseInfo.Builder.newBuilder()
```
I also created integration test for it. The case is put into StaticClientEndpointsTest.java. And to facilitate the integration test, I also modified staticclient.yml file, add a service with mixed case service Id

Signed-off-by: zhuxun <zhuxun@cn.ibm.com>

JirkaAichler commented:
Looks great!. However catalog also needs to be updated to support mixed case service ID. When I setup Api Doc, I see this: 

![image](https://user-images.githubusercontent.com/39735421/56899117-f59d3180-6a92-11e9-841c-9e6359c03196.png)

JirkaAichler commented:
Mainframe integration tests are failing because they are using the different version of 'config/local/api-defs/staticclient.yml' stored in Broadcom private repository. I will update this file once we will resolve the problem with the catalog.

arxioly commented on gateway-service/src/main/java/com/ca/mfaas/gateway/routing/MfaasRouteLocator.java:
```diff
@@ -86,7 +87,20 @@
                 }
-                List<String> keys = createRouteKeys(serviceInstances, routedServices, serviceId);
+
+                //serviceId is converted to lowercase by Discovery Service. Recover serviceId which contains uppercase characters
```
Could you, please, move this block to separate method? Like this:
```java
    private String getOriginalServiceId(String serviceId, List<ServiceInstance> serviceInstances) {
        //serviceId is converted to lowercase by Discovery Service. Recover serviceId which contains uppercase characters
        String originalServiceId = serviceId;
        if (serviceInstances.size() > 0) {
            ServiceInstance instance = serviceInstances.get(0);
            if (instance instanceof EurekaDiscoveryClient.EurekaServiceInstance) {
                String vipAddr = ((EurekaDiscoveryClient.EurekaServiceInstance) instance).getInstanceInfo().getVIPAddress();
                if (serviceId.equalsIgnoreCase(vipAddr)) {
                    originalServiceId = vipAddr;
                }
            }
        }
        return originalServiceId;
    }
```
And put the calling of it directly in to calling of `createRouteKeys` method, like this: 
```java
List<String> keys = createRouteKeys(serviceInstances, routedServices, getOriginalServiceId(serviceId, serviceInstances));
```

CindyZX replied:
Thank you for the comments, I've pushed a commit, and modified the code as your comments, thank you

arxioly commented on integration-tests/src/test/java/com/ca/mfaas/discoveryservice/staticdef/StaticClientEndpointsTest.java:
```diff
@@ -34,4 +35,14 @@ public void shouldSameResponseAsDynamicService() throws Exception {
+     */
+    @Test
+    public void shouldMixedCaseServiceIdWork() throws Exception {
+        final HttpResponse staticResponse = HttpRequestUtils.getResponse(MIXED_CASE_GREET, SC_OK);
```
Please, don't need to create a variable, if you don't use it. Thanks.

CindyZX replied:
Thank you for the comments, I've modified the code in latest commit

CindyZX commented:
Thank you @JirkaAichler , I just pushed a commit, in which I updated Catalog to support mixed case serviceId, I also modified some code according to review comments in the commit

Here is how the updated catalog looks like in my local machine
![image](https://user-images.githubusercontent.com/10337143/56947345-70cc1980-6b5f-11e9-9732-d76a8eeab76d.png)

JirkaAichler commented:
Please could you also fix the service title?

![image](https://user-images.githubusercontent.com/39735421/56960075-efb95600-6b4f-11e9-8e5a-48148bc585b4.png)

Thanks!

JirkaAichler commented:
I think this change also requires the documentation update.
See for example: https://zowe.github.io/docs-site/latest/extend/extend-apiml/api-mediation-onboard-an-existing-rest-api-service-without-code-changes.html#configuration-parameters

This generated from: https://github.com/zowe/docs-site/blob/docs-staging/docs/extend/extend-apiml/api-mediation-onboard-an-existing-rest-api-service-without-code-changes.md
Another PR for doc-site will be needed,

There are more pages and files mentioning to use lower case only. Our team can update the doc if you have troubles to do it.

JirkaAichler commented:
Cindy, can we discuss the solution on Tuesday after Zowe sync up meeting? We'd like to see your use case and we will try to come up with the better solution.

We would like to rely on serviceId instead of VIPAddress. It would create a confusion in our code..

CindyZX commented:
Hi Jirka, sure, we can discuss it after Tuesday Zowe sync-up meeting

CindyZX commented:
Close this PR since it is against APIML conformance criteria, and it will also bring up other problems
